pennylane_lightning_kokkos/src/algorithms/ObservablesKokkos.hpp

```diff
@@ -0,0 +1,504 @@
+            new TensorProdObsKokkos(std::move(obs))};
+    }
+
+    static auto create(std::vector<std::shared_ptr<ObservableKokkos<T>>> obs)
+        -> std::shared_ptr<TensorProdObsKokkos<T>> {
+        return std::shared_ptr<TensorProdObsKokkos<T>>{
+            new TensorProdObsKokkos(std::move(obs))};
+    }
+
+    /**
+     * @brief Get the number of operations in observable.
+     *
+     * @return size_t
+     */
+    [[nodiscard]] auto getSize() const -> size_t { return obs_.size(); }
+
+    /**
+     * @brief Get the wires for each observable operation.
+     *
+     * @return const std::vector<std::vector<size_t>>&
+     */
+    [[nodiscard]] auto getWires() const -> std::vector<size_t> override {
+        return all_wires_;
+    }
+
+    void applyInPlace(StateVectorKokkos<T> &sv) const override {
+        for (const auto &ob : obs_) {
+            ob->applyInPlace(sv);
+        }
+    }
+
+    [[nodiscard]] auto getObsName() const -> std::string override {
+        using Pennylane::Util::operator<<;
+        std::ostringstream obs_stream;
+        const auto obs_size = obs_.size();
+        for (size_t idx = 0; idx < obs_size; idx++) {
+            obs_stream << obs_[idx]->getObsName();
+            if (idx != obs_size - 1) {
+                obs_stream << " @ ";
+            }
+        }
+        return obs_stream.str();
+    }
+};
+
+/**
+ * @brief General Hamiltonian as a sum of observables.
+ *
+ */
+template <typename T>
+class HamiltonianKokkos final : public ObservableKokkos<T> {
+  public:
+    using PrecisionT = T;
+
+  private:
+    std::vector<T> coeffs_;
+    std::vector<std::shared_ptr<ObservableKokkos<T>>> obs_;
+
+    [[nodiscard]] bool
+    isEqual(const ObservableKokkos<T> &other) const override {
+        const auto &other_cast =
+            static_cast<const HamiltonianKokkos<T> &>(other);
+
+        if (coeffs_ != other_cast.coeffs_) {
+            return false;
+        }
+
+        for (size_t i = 0; i < obs_.size(); i++) {
+            if (*obs_[i] != *other_cast.obs_[i]) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+  public:
+    /**
+     * @brief Create a Hamiltonian from coefficients and observables
+     *
+     * @param arg1 Arguments to construct coefficients
+     * @param arg2 Arguments to construct observables
+     */
+    template <typename T1, typename T2>
+    HamiltonianKokkos(T1 &&arg1, T2 &&arg2)
+        : coeffs_{std::forward<T1>(arg1)}, obs_{std::forward<T2>(arg2)} {
+        assert(coeffs_.size() == obs_.size());
+    }
+
+    /**
+     * @brief Convenient wrapper for the constructor as the constructor does not
+     * convert the std::shared_ptr with a derived class correctly.
+     *
+     * This function is useful as std::make_shared does not handle
+     * brace-enclosed initializer list correctly.
+     *
+     * @param arg1 Argument to construct coefficients
+     * @param arg2 Argument to construct terms
```

Comment:
Would be nice to have parameters with meaningful names.

Reply:
sure!

Reply:
done